### PR TITLE
Kicks players after being AFK too long

### LINKED ISSUE #10577: Need AFK kick timer
People are sitting in rounds doing nothing to pad their role times and taking up a slot.

AFKSystem issues events so all you need to do is have some kind of timer that accumulates for AFK people (e.g. 5-10 minutes) and kicks them after sending a warning.

### PR DESCRIPTION
**Pull request by tomasalves8**
## About the PR 
- Creates a CVar with the amount of time before an AFK player is kicked. (Default is 10 minutes)
- One minute before the player is kicked, he's sent a warning.

- Adds `colorOverride` to `DispatchServerMessage`.

Fixes #10577 

**Screenshots**

**Changelog**

:cl:
- tweak: Kick AFK Players

### CODE REVIEW

File: Content.Server/Afk/AFKSystem.cs [hunk 1]
```diff
@@ -18,23 +20,30 @@ public sealed class AFKSystem : EntitySystem
...
+        _configManager.OnValueChanged(CCVars.AfkKickTime, SetAfkKickDelay, true);
     }
 
     private void SetAfkDelay(float obj)
     {
         _checkDelay = obj;
     }
+    private void SetAfkKickDelay(int obj)
+    {
+        _kickDelay = obj;
```
Comment (metalgearsloth) [inline Content.Server/Afk/AFKSystem.cs:45]: Setting a float to an int?

File: Content.Server/Afk/AFKSystem.cs [hunk 2]
```diff
@@ -77,16 +86,30 @@ public override void Update(float frameTime)
...
             var pSession = (IPlayerSession) session;
             var isAfk = _afkManager.IsAfk(pSession);
-
-            if (isAfk && _afkPlayers.Add(pSession))
+            if (isAfk)
             {
-                var ev = new AFKEvent(pSession);
-                RaiseLocalEvent(ref ev);
-                continue;
+                var curTime = DateTimeOffset.Now.ToUnixTimeSeconds();
```
Comment (metalgearsloth) [inline Content.Server/Afk/AFKSystem.cs:91]: You should be using IGameTiming to get times.

File: Content.Server/Afk/AFKSystem.cs [hunk 3]
```diff
@@ -77,16 +86,30 @@ public override void Update(float frameTime)
...
+                if (!_afkPlayers.ContainsKey(pSession))
+                {
+                    _afkPlayers.Add(pSession, curTime);
+                    var ev = new AFKEvent(pSession);
+                    RaiseLocalEvent(ref ev);
+                    continue;
+                }
+                var timeElapsed = curTime - _afkPlayers[pSession];
+                if (timeElapsed > _kickDelay*60)
```
Comment (metalgearsloth) [inline Content.Server/Afk/AFKSystem.cs:100]: Just use the float directly rather than multiplying?

File: Content.Server/Afk/AFKSystem.cs [hunk 4]
```diff
@@ -77,16 +86,30 @@ public override void Update(float frameTime)
...
+                {
+                    var network = IoCManager.Resolve<IServerNetManager>();
```
Comment (metalgearsloth) [inline Content.Server/Afk/AFKSystem.cs:102]: This should be cached as a dependency.

File: Content.Shared/CCVar/CCVars.cs [hunk 5]
```diff
@@ -1029,6 +1029,12 @@ public sealed class CCVars : CVars
         public static readonly CVarDef<float> AfkTime =
             CVarDef.Create("afk.time", 60f, CVar.SERVERONLY);
 
+        /// <summary>
+        /// How long a client can go after being detected as AFK before being kicked;
+        /// </summary>
+        public static readonly CVarDef<int> AfkKickTime =
+            CVarDef.Create("afk.kick_time", 10, CVar.SERVERONLY);
```
Comment (metalgearsloth) [inline Content.Shared/CCVar/CCVars.cs:1036]: This should be a float and in seconds.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Afk/AFKSystem.cs [hunk 6]
```diff
@@ -18,23 +21,32 @@ public sealed class AFKSystem : EntitySystem
...
+    [Dependency] private readonly IChatManager _chatManager = default!;
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
+    [Dependency] private readonly IServerNetManager _netManager = default!;
 
     private float _checkDelay;
+    private float _kickDelay;
     private float _accumulator;
 
-    private readonly HashSet<IPlayerSession> _afkPlayers = new();
+    private readonly Dictionary<IPlayerSession, double> _afkPlayers = new();
```
Comment (mirrorcult) [inline Content.Server/Afk/AFKSystem.cs:32]: why is this a double when its a float everywhere else

File: Content.Server/Afk/AFKSystem.cs [hunk 7]
```diff
@@ -77,16 +89,29 @@ public override void Update(float frameTime)
...
+                    var ev = new AFKEvent(pSession);
+                    RaiseLocalEvent(ref ev);
+                    continue;
+                }
+                var timeElapsed = curTime - _afkPlayers[pSession];
+                if (timeElapsed > _kickDelay)
+                {
+                    _netManager.DisconnectChannel(pSession.ConnectedClient, Loc.GetString("afk-system-kick-reason"));
+                }
+                else if(timeElapsed >= _kickDelay-1)
```
Comment (mirrorcult) [inline Content.Server/Afk/AFKSystem.cs:107]: isnt this going to give the warning 1 second before you're kicked?

**Review by mirrorcult — CHANGES_REQUESTED** [review]
admins should be exempt from AFK kicking--I often leave the game in the background so I can respond to ahelps

File: Content.Server/Afk/AFKSystem.cs (hunk 7 above)
Comment (metalgearsloth) [reply Content.Server/Afk/AFKSystem.cs:107]: Also it's going to send it every time the afk system runs; if a downstream has this at 1 second intervals the message is gonna get spammed a lot.

File: Content.Server/Afk/AFKSystem.cs [hunk 8]
```diff
@@ -77,16 +89,29 @@ public override void Update(float frameTime)
...
+            if (isAfk)
             {
-                var ev = new AFKEvent(pSession);
-                RaiseLocalEvent(ref ev);
-                continue;
+                var curTime = _gameTiming.CurTime.TotalSeconds;
+                if (!_afkPlayers.ContainsKey(pSession))
+                {
+                    _afkPlayers.Add(pSession, curTime);
+                    var ev = new AFKEvent(pSession);
```
Comment (metalgearsloth) [inline Content.Server/Afk/AFKSystem.cs:98]: TryAdd?

File: Content.Server/Afk/AFKSystem.cs [hunk 9]
```diff
@@ -77,16 +89,29 @@ public override void Update(float frameTime)
...
+                else if(timeElapsed >= _kickDelay-1)
+                {
+                    _chatManager.DispatchServerMessage(pSession, Loc.GetString("afk-system-kick-warning"), Color.Red);
+                }
             }
-
-            if (!isAfk && _afkPlayers.Remove(pSession))
+            else if (_afkPlayers.ContainsKey(pSession))
             {
+                _afkPlayers.Remove(pSession);
```
Comment (metalgearsloth) [inline Content.Server/Afk/AFKSystem.cs:114]: Remove returns a bool.

**Pull request closed without merging**